### ISSUE TITLE
EC2 Metadata Client

### ISSUE THREAD

**Issue by davidmankin** [aws/aws-sdk-java-v2#61]
Since all access to EC2MetadataUtils is static, the class is very difficult to use in a properly tested application.  It can't be mocked; you can't give canned answers when running in a non-EC2 environment.

If there were an EC2Metadata interface or abstract class of which there were a default implementation, then it would be easier to use.  

A contrived example:
```java
public class Application {
    EC2Metadata metadata;
    public static void main(String args[]) {
        new Application(EC2MetadataUtils.getDefaultInstance());
    }
    public Application(EC2Metadata metadata) {
        this.metadata = metadata;
        LOG.info("We're running on " + metadata.getInstanceId());
    }
}

public class ApplicationTest {
    @Mock EC2Metadata metadata;
    @Test public testApp() {
        when(metadata.getDefaultInstance()).then(return "i-fakeinstance");
        new Application(metadata);
        verifyLoggedLine("We're running on i-fakeinstance");
    }
}
```

As it stands today, I need to run a whole HTTP process that mimics the metadata server and somehow get the system property to point to it; and even that is a global setting for the whole JVM: parallel tests can interfere with each other.  Or I can write my own EC2Metadata interface and use it throughout my app, but it's not something each developer should have to do.

**Comment by millems** [comment]
Using this issue to track building a public-facing EC2 Metadata Client.

**Comment by dagnir** [comment]
Sorry for the delay in our response. We aim to respond to all comments, feedback and pull requests our customers make and we apologize that we failed to do so in this instance.

We have been working with the EC2 team to figure out the best possible customer experience for exposing this information, and in the meantime we welcome all suggestions and will use them to help us in that decision making process.

We will keep updating this issue as we move towards delivery.

Thank you for your patience!

**Comment by injeniero** [comment]
@dagnir Any update on this?

**Comment by spender0** [comment]
any progress? It turned out that we need to change metadata_service_timeout as the connection in our case takes more than 1 second. But it is hardcoded in the SDK

**Comment by sanjeeb26** [comment]
@dagnir do you have timeline when this will be available in the SDK?

**Comment by dFarras** [comment]
@millems @dagnir is there any new about this issue?

Thanks in advance

**Comment by millems** [comment]
@dFarras Sorry, not right now. What specific feature do you need (and why do you need it), so we can make sure that's included when this gets done?

**Comment by dFarras** [comment]
My use case is that I need to retrieve an instance id but I can not be sure were the instance is deployed in advance. 

But instance id is only informative info and therefore no user will desire to wait for too long. 
After diving a bit in Ec2Metadata I saw no obvious way of injecting an http client or modifying its configuration so when I call its methods they will fail after http standard timeout which is to long for me.

I have a workaround, but I would desire to be able to modify to modify Ec2Metadata configuration.

**Comment by millems** [comment]
Feature request from a customer: Once this is implemented, it would be nice to be able to configure which client is used by the instance metadata credentials provider, so that things like the endpoint or timeouts can be configured.

**Comment by L-Applin** [comment]
The EC2 Metadata Client now available, starting in version 2.19.29
https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/imds/Ec2MetadataClient.html
